### ISSUE TITLE
v2

### ISSUE THREAD

**Issue by zernonia** [unovue/reka-ui#908]
# Announcement: Changes Coming to Radix Vue

## Introduction
First and foremost, we really want to express appreciation for all the effort put in by the [Radix UI](https://github.com/radix-ui) team in creating the absolutely powerful, unstyled, high-quality, and accessible components that this project was originally ported from.

## Background
However, as feature requests grew, Radix Vue began to add more components and features, and modify certain behaviors specifically tailored for Vue. As a result, Radix Vue started to feel like a unique component library rather than just the ported version of Radix UI. Therefore, we are planning to give it a different identity.

## What's Coming in v2?
- Brand change
- NPM registry change
- Some component refactoring [link to project](https://github.com/orgs/unovue/projects/3) https://github.com/radix-vue/radix-vue/labels/v2

## Note
1. We will still be pushing bug fixes and introducing some new primitive components prior to the v2 release.
2. We are still brainstorming the new identity name. Please share any ideas you have in the comments below.
3. We are open to suggestions for areas of improvement. Please ensure that any suggestions are sensible and would result in a general improvement rather than addressing only a specific edge-case.

> [!IMPORTANT]
> We understand that these changes may require some major adjustments in your project, and we apologize for any inconvenience in advance.

**Comment by khairulhaaziq** [comment]
Hi @zernonia ! Big decision! I had my concerns but after several thought and looking back at the project, I think a rebrand is a really good idea! I think what you and other contributors have been building and currently building has turned this library into a very big project, and in many sense it has grown larger than the radix primitives. I think a rebrand is sign of growth and by detaching from the previous radix brand, it removes a significant blocker, allowing decisions to be made without the concern of fitting the radix brand. Overall i'm very excited for the future of the library and see what you'll come up with 🚀🚀🚀

**Comment by gustojs** [comment]
How about something that works both ways? Can be regarded as something separate, but still points at the origins. Vuedix, for example.

**Comment by MellKam** [comment]
@zernonia For a long time, I had an idea to create my own UI library and call it `unbound-ui`. I even reserved the package name and scope for it on npm, but I never had the chance to really start the work. If you like the name, I can transfer the ownership to you. I think it nicely represents the purpose of this library and hasn't been used in other projects yet.

**Comment by edtorba** [comment]
Just a hint, you can also look into Latin words, at least it seems that's what the Radix team did, e.g., radix = root (it seems). What you want is something simple and catchy, while also being unique."

**Comment by Rolanddoda** [comment]
@zernonia does this means that there are going to be breaking changes in v2? 
I think there might be a lot of people evaluating Radix Vue and if 1.x is going to be stale after a few months, then that should be made clear.

**Comment by cyyynthia** [comment]
While the project doesn't follow semver, a major version bump is always indicative of some breaking changes. I don't think the project has a clear list of what's going to break, but form-related components are planned to go through changes such as https://github.com/radix-vue/radix-vue/issues/1017, internal changes that may affect their behavior (backing hidden field, exposed API, ...)

Radix Vue published breaking changes *for components in alpha* in patch-releases, and these patch-releases often contain features.

As someone who recently adopted Radix Vue, I can say for those looking to adopt it it's very promising and we're having great results in the project it's being used in, but don't underestimate the alpha label on components (see prev. remark on breaking changes), and remember that the project is quite young so there are a handful of areas where it can improve 😅

I don't think v2 will bring huge breaking changes, but I'm not a maintainer of the project and haven't been in its internals for long enough to know what needs to change. So long your project follows a healthy component structure (you generally shouldn't use raw components and wrap them with your own styles first; or use something like shadcn-vue which is versioned separately) the changes shouldn't be too bad - but, I'm not a maintainer just a user who really enjoyed using this lib thus far 😃 

Possibly relevant issues tagged https://github.com/radix-vue/radix-vue/labels/v2:
- https://github.com/radix-vue/radix-vue/issues/1046
- https://github.com/radix-vue/radix-vue/issues/1043
- https://github.com/radix-vue/radix-vue/issues/1017
- https://github.com/radix-vue/radix-vue/issues/498

**Comment by zernonia** [comment]
@Rolanddoda  Yes in the description we mention that v2 will have breaking changes (major adjustments in your project). Worry not that we will mentioned all of that breaking changes in the changelog, and will be releasing `rc`'s earlier too.

@cyyynthia Currently we try our best to follow semver 😁, especially for the breaking changes part. In minor version we have new changes but with deprecation warning, unless for `alpha` component. But we try to minimize the breaking in `alpha` as well and aimed to publish major changes in the same minor version.

**Comment by teleskop150750** [comment]
Расписал по пунктам чем Perigee-ui отличается от radix-vue

1) I use let variables where React uses useRef. Radix-vue [creates unnecessary reactive ref](https://github.com/radix-vue/radix-vue/blob/3f0f965fcf6fc3901e4fbbedf9a68dcb7d706f3f/packages/radix-vue/src/HoverCard/HoverCardRoot.vue#L64) variables in all places, which is completely unnecessary.

2) I use the [composeEventHandlers hook](https://github.com/radix-ui/primitives/blob/660060a765634e9cc7bf4513f41e8dabc9824d74/packages/core/primitive/src/primitive.tsx#L1). This hook allows canceling events through preventDefault. Unfortunately, in Vue, [handlers received from the parent component end up at the end of the array and are called last](https://github.com/vuejs/core-vapor/blob/30583b9ee1c696d3cb836f0bfd969793e57e849d/packages/runtime-core/src/vnode.ts#L886) (I might have misunderstood the code, there may be inaccuracies).

3) I use attribute inheritance instead of redefining them in each component, [generating unnecessary code](https://github.com/radix-vue/radix-vue/blob/3f0f965fcf6fc3901e4fbbedf9a68dcb7d706f3f/packages/radix-vue/src/shared/useForwardProps.ts#L16). Moreover, it seems that Volar now[ has typing for attrs](https://github.com/vuejs/language-tools/pull/4103). This approach may change if it turns out to be inconvenient to use.

4) I do not use their [useForwardExpose](https://github.com/radix-vue/radix-vue/blob/3f0f965fcf6fc3901e4fbbedf9a68dcb7d706f3f/packages/radix-vue/src/shared/useForwardExpose.ts#L21). A hook that replaces the original expose object to pass props outside. Why, if access to them is [already available](https://vuejs.org/api/component-instance.html#props).

5) I do not use asChild for implementing primitives. Instead, I wrap the component's [content in a hook and use it](https://github.com/perigee-ui/vue-primitives/blob/a991db71fbecf364cd0b8479b294606236b104b4/packages/vue-primitives/src/dialog/DialogContentModal.vue#L65). I thought it was a bit cumbersome to use so many empty wrapper components when they can be eliminated. For example, the FocusScope wrapper is 3 components: FocusScope -> Primitive -> Slot. If there are 3 such wrappers, it is already a tree of 9+ components. This is currently a test implementation.

6) I think it's unlikely. In the distant future, there is a thought to abandon components and leave only hooks that return props. For example, like in [Zag](https://zagjs.com/components/react/accordion) and [Melt](https://melt-ui.com/docs/introduction). If the Vapor mod comes out in our lifetime, this method will allow us to get rid of component is and [modification of children](https://github.com/radix-vue/radix-vue/blob/3f0f965fcf6fc3901e4fbbedf9a68dcb7d706f3f/packages/radix-vue/src/Primitive/Slot.ts#L12) for asChild.

7) Different implementation of Collection.

8) Presence is just a [hook for me](https://github.com/perigee-ui/vue-primitives/blob/7c341db59fdfdb0cc88dfa6614d6c390b6856780/packages/vue-primitives/src/presence/usePresence.ts#L8). The same approach was found in Solid's Kobalte. In radix-vue, it is the reason for the absence of content on the first render in components like [Accordion](https://github.com/radix-vue/radix-vue/issues/978).

9) Availability of component context export for access in user code. It seems that this was promised to be added in Radix v2. I am not following.

10) No [empty wrappers that do nothing](https://github.com/radix-vue/radix-vue/blob/main/packages/radix-vue/src/AlertDialog/AlertDialogTrigger.vue).

**Comment by yacinehmito** [comment]
@teleskop150750 Why are you outlining the differences between your library and this one in the present issue? I don't see anyone asking the question here. It feels inappropriate.

**Comment by teleskop150750** [comment]
@yacinehmito Where is it better to write to discuss the options for cooperation

**Comment by teleskop150750** [comment]
@yacinehmito I described what I wanted to see in radix 2. Where is it better to write to discuss the options for cooperation?

**Comment by zernonia** [comment]
Thanks for sharing your thoughts @teleskop150750  .

While I appreciate the efforts to streamline internal structures, I believe that most of these suggestions mainly target internal component architecture rather than offering tangible benefits to the end-user. The primary goal of v2 is on features that require breaking changes and is designed to bring tangible improvements for the end user.

Some of your ideas could be valuable for long-term architectural improvements, but prioritizing internal changes without direct developer-facing advantages seems like a waste of effort.

With that said, looking forward to seeing any benchmarks where these changes benefit users directly tho 😉  Feel free to post them on discussion page

**Comment by sadeghbarati** [comment]
@teleskop150750 Hi 👋 

Each of the list items you described can be a PR and tested with [`pkg.pr.new`](https://github.com/radix-vue/radix-vue/pull/1227#issuecomment-2295657036) capability, 
You are free to research creating your own lib for your use cases.

But **contributing**/**improving** one project that is used by thousands of people is way better imo

**Comment by teleskop150750** [comment]
> But **contributing**/**improving** one project that is used by thousands of people is way better imo

The first such project was OkuUI. I have already promised its author to help complete it and implement some of the listed ideas. It is a pity that Radix Vue did not want to cooperate and join forces and made its own version of the library

**Comment by teleskop150750** [comment]
Is it possible to place interactive elements inside other interactive elements? [Input inside Button](https://github.com/unovue/radix-vue/blob/5923754fead604934c4b9f92a542665fcc759f39/packages/core/src/Checkbox/CheckboxRoot.vue#L143)

[Html Can Include](https://caninclude.glitch.me/caninclude?child=input&parent=button)
[spec](https://html.spec.whatwg.org/#interactive-content-2)

**Comment by teleskop150750** [comment]
Is it possible to place interactive elements inside other interactive elements? [Input inside Button](https://github.com/unovue/radix-vue/blob/5923754fead604934c4b9f92a542665fcc759f39/packages/core/src/Checkbox/CheckboxRoot.vue#L143)

[Html can iclude](https://caninclude.glitch.me/caninclude?child=input&parent=button)

[spec](https://html.spec.whatwg.org/#interactive-content-2)

**Comment by teleskop150750** [comment]
Is it possible to place interactive elements inside interactive elements? [Input inside the button
](https://github.com/unovue/radix-vue/blob/5923754fead604934c4b9f92a542665fcc759f39/packages/core/src/Checkbox/CheckboxRoot.vue#L143)

[Html can include](https://caninclude.glitch.me/caninclude?child=input&parent=button)

![image](https://github.com/user-attachments/assets/e3243eee-b86b-4b7d-9e1c-1932d38e6187)

**Comment by iamandrewluca** [comment]
@zernonia, in the issue description, the link to the project seems to be incorrect

https://github.com/orgs/unovue/projects/3

**Comment by mary-ext** [comment]
thoughts on exposing components like FocusScope, FocusGuards, and DismissableLayer as public API?

I get that Radix (React) has them as internal APIs as well but we're currently having problems at work adapting Dialog due to their clunky design, so at the very least we want to at the very least have some consistency with other Radix components we are using.

**Comment by orenmizr** [comment]
i am building something right now with shadcn-vue. which depends on radix-vue. they are great. i am concerened about too many breaking changes.  we're tallking about bring your own comoponent system... not npm package - sounds like it going to be painful. 

also. a name is important. i used the radix popularity to explain me leaning into radix based systems. as mentioned, i like vue projects that are clerverly name to show the connection to another framework (nuxt/next). 

btw, Reka means background in my language.

**Comment by mkeyy0** [comment]
Regarding the name, I don't have a specific proposal. But I agree with @gustojs proposal. It's better to keep some kind of reference that this library was created using Radix as a reference. The reason is I read some discussions on Reddit and other forums many people see the radix library in React and are searching for an alternative in a Vue world. And that's how I initially found this great library. I've just searched Radix with vue and found this library. I guess some people found this library in this way as well.

**Comment by mkeyy0** [comment]
Also, I have a question regarding breaking changes. I'm using the shadcn vue library with Radix as well. I don't know if this question is to you specifically, but are you considering that shadcn heavily relies on Radix components during the development? Because it would be nice to keep the connection between those 2 libraries.

**Comment by teleskop150750** [comment]
> Thanks for sharing your thoughts @teleskop150750 .
> 
> While I appreciate the efforts to streamline internal structures, I believe that most of these suggestions mainly target internal component architecture rather than offering tangible benefits to the end-user. The primary goal of v2 is on features that require breaking changes and is designed to bring tangible improvements for the end user.
> 
> Some of your ideas could be valuable for long-term architectural improvements, but prioritizing internal changes without direct developer-facing advantages seems like a waste of effort.
> 
> With that said, looking forward to seeing any benchmarks where these changes benefit users directly tho 😉 Feel free to post them on discussion page

I tried to check the performance beforehand. There may be errors in the code, so don't trust the results too much.

![image](https://github.com/user-attachments/assets/1ddefdc1-381a-4f82-9a67-9017941e2a86)
![image](https://github.com/user-attachments/assets/84d6182e-036b-498e-9729-ab565478a0ed)

**Comment by teleskop150750** [comment]
> Thanks for sharing your thoughts @teleskop150750 .
> 
> While I appreciate the efforts to streamline internal structures, I believe that most of these suggestions mainly target internal component architecture rather than offering tangible benefits to the end-user. The primary goal of v2 is on features that require breaking changes and is designed to bring tangible improvements for the end user.
> 
> Some of your ideas could be valuable for long-term architectural improvements, but prioritizing internal changes without direct developer-facing advantages seems like a waste of effort.
> 
> With that said, looking forward to seeing any benchmarks where these changes benefit users directly tho 😉 Feel free to post them on discussion page

I tried to [check the performance](https://github.com/oku-ui/oku-benchmark/tree/oku) beforehand. There may be errors in the code, so don't trust the results too much.

![IMG_20241122_023506_033](https://github.com/user-attachments/assets/01f8a201-20f6-466e-b7e2-7e363cf8aed6)
![IMG_20241122_023506_028](https://github.com/user-attachments/assets/5845f9f3-62d9-42c1-96bb-ab4de20e83af)
![IMG_20241122_023505_596](https://github.com/user-attachments/assets/92a822a0-6a54-443d-880a-3e11488fe54f)

**Comment by zernonia** [comment]
> Also, I have a question regarding breaking changes. I'm using the shadcn vue library with Radix as well. I don't know if this question is to you specifically, but are you considering that shadcn heavily relies on Radix components during the development? Because it would be nice to keep the connection between those 2 libraries.

Yeah we are working on updated version of shadcn-vue with `reka-ui`. Once that is ready we will release both together 😁

**Comment by jd-solanki** [comment]
> we are working on updated version of shadcn-vue with reka-ui

- I think this will be based on TW v4 beta right?
- Is we're rebarding shadcn-vue to something else?
- I'm just starting new project with shadcn-vue so can we start with new alpha or current stable?

**Comment by romansp** [comment]
VueUse released v12 with Vue 3 support only. One could dream about bumping it as well in v2. 

I guess we could even do this in v1 as Vue 2 was never supported by radix-vue afaik?

**Comment by zernonia** [comment]
Yooo @jd-solanki !

> * I think this will be based on TW v4 beta right?

I don't think this is in shadcn's roadmap for now. But we shall see 😁


> * Is we're rebarding shadcn-vue to something else?

Nope, shadcn-vue will stays, but we will have `reka-ui` version of shadcn-vue hosted under `@next` tag 
https://next.shadcn-vue.com/


> * I'm just starting new project with shadcn-vue so can we start with new alpha or current stable?


Currently `reka-ui` still in `alpha`, so you should use `shadcn-vue` for now, and the upgrade to `next.shadcn-vue` should be relatively straightforward by just running `npx shadcn-vue@next add <needed component>` when it's ready 😁

**Comment by HummingMind** [comment]
> Thanks @romansp ! That sounds like an amazing news! We can include this upgrade in v2 😁
> 
> And yes, radix-vue only support Vue 3, so we can technically upgrade it as well. But since it's a major version upgrade, I afraid that would incur some unforeseen breaking changes, so best not to upgrade first 😬

It's a non-breaking release, and is equivalant to v11.3. The major version bump was to drop Vue2 support. I don't think anything else changed.

**Comment by zernonia** [comment]
@HummingMind  correct. I mention breaking change because `radix-vue` still using `vueuse@10.11.0` 😁

**Comment by sawden** [comment]
Please consider releasing a v2 alpa/beta to npm soon 🙏 
Alternatively, adding <NavigationMenuRoot/> `disableHoverTrigger` to the next v1 release would be greatly appreciated, as it introduces no breaking changes ([Commit](https://github.com/unovue/radix-vue/commit/4b34df6dc9470b1df74725cb76eb63b7d7f128fc))

Thank you for your work! 🙂

**Comment by Radiergummi** [comment]
> Alternatively, adding disableHoverTrigger to the next v1 release would be greatly appreciated, as it introduces no breaking changes ([Commit](https://github.com/unovue/radix-vue/commit/4b34df6dc9470b1df74725cb76eb63b7d7f128fc))

Seconded, even if I seem to be ignored by everyone around here; this would be super helpful.

**Comment by ilyaliao** [comment]
> VueUse released v12 with Vue 3 support only. One could dream about bumping it as well in v2.
> 
> I guess we could even do this in v1 as Vue 2 was never supported by radix-vue afaik?

Correct, the `v12` version mainly drops support for `Vue 2`, but there have also been fixes for some issues in the versions after v12, so I think you can consider upgrading to the latest version :)
